### PR TITLE
support PackedSequence type for apply_for_tensors

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76265

support PackedSequence type for apply_for_tensors, some rnn modules outputs are PackedSequence types

Differential Revision: [D35862156](https://our.internmc.facebook.com/intern/diff/D35862156/)

### CODE REVIEW

File: torch/distributed/fsdp/utils.py [hunk 1]
```diff
@@ -8,18 +9,21 @@
 
 
 def _apply_to_tensors(
-    fn: Callable, container: Union[torch.Tensor, Dict, List, Tuple, Set]
+    fn: Callable, container: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict, PackedSequence]
 ) -> Any:
     """Recursively apply to all tensor in different kinds of container types."""
 
-    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict]) -> Any:
+    def apply(x: Union[torch.Tensor, Dict, List, Tuple, Set, OrderedDict, PackedSequence]) -> Any:
```
Comment (mrshenli) [inline torch/distributed/fsdp/utils.py:16]: curious, do we need to support custom types (e.g., by allowing users to pass in a function or by pickling `x` if it is none of the listed types here?)?

File: test/distributed/fsdp/test_utils.py [hunk 2]
```diff
@@ -92,6 +93,23 @@ def test_replace_by_prefix(self):
...
+        seq_length = torch.tensor([4], dtype=torch.int)
+
+        def fill_fn(x):
+            x.fill_(0)
+
+        x = nn.utils.rnn.pack_padded_sequence(x, seq_length)
+        x, h = rnn(x)
+        x = _apply_to_tensors(fill_fn, x)
+        x, _ = nn.utils.rnn.pad_packed_sequence(x)
+        assert torch.sum(x) == 0
```
Comment (mrshenli) [inline test/distributed/fsdp/test_utils.py:110]: any reason we are using `assert` instead of `self.assertEqual`?

**Review by mrshenli — APPROVED** [review]

**Review by rohan-varma — APPROVED** [review]
LG

File: torch/distributed/fsdp/utils.py (hunk 1 above)
Comment (zhaojuanmao) [reply torch/distributed/fsdp/utils.py:16]: the apply_to_tensors is mainly used for casting input tensors to mixed precision or register hooks of output tensors internally inside FSDP. possibly it is hard to expose to users for passing a function or so. We also do not want to pickling X for non-tensor data, e..g, we do not want to cast scalars or register hooks of scalars. 

but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

Comment (mrshenli) [reply torch/distributed/fsdp/utils.py:16]: > but I'm curious how PyTorch generally get tensors from a custom defined type...? if users' input or output is a custom defined type and it contains tensors, we may not "apply_to_tensors" as expected for this custom defined type right now

FWIW, there is no good solution today. What CUDA Future does is pickling the Python obj and extract all tensors during the pickling process, but that can be pretty slow.

**Pull request closed without merging**